man/mancxx.roff/Db.3
.ds TYPE CXX
.\"
.\" See the file LICENSE for redistribution information.
.\"
.\" Copyright (c) 1997
.\"	Sleepycat Software.  All rights reserved.
.\"
.\"	@(#)Db.sox	10.11 (Sleepycat) 1/18/98
.\"
.\"
.\" See the file LICENSE for redistribution information.
.\"
.\" Copyright (c) 1997
.\"	Sleepycat Software.  All rights reserved.
.\"
.\"	@(#)macros.so	10.39 (Sleepycat) 1/22/98
.\"
.\" We don't want hyphenation for any HTML documents.
.ie '\*[HTML]'YES'\{\
.nh
\}
.el\{\
.ds Hy
.hy
..
.ds Nh
.nh
..
\}
.\" The alternative text macro
.\" This macro takes two arguments:
.\"	+ the text produced if this is a "C" manpage
.\"	+ the text produced if this is a "CXX" or "JAVA" manpage
.\"
.de Al
.ie '\*[TYPE]'C'\{\\$1
\}
.el\{\\$2
\}
..
.\" Scoped name macro.
.\" Produces a_b, a::b, a.b depending on language
.\" This macro takes two arguments:
.\"	+ the class or prefix (without underscore)
.\"	+ the name within the class or following the prefix
.de Sc
.ie '\*[TYPE]'C'\{\\$1_\\$2
\}
.el\{\
.ie '\*[TYPE]'CXX'\{\\$1::\\$2
\}
.el\{\\$1.\\$2
\}
\}
..
.\" Scoped name for Java.
.\" Produces Db.b, for Java, otherwise just b.  This macro is used for
.\" constants that must be scoped in Java, but are global otherwise.
.\" This macro takes two arguments:
.\"	+ the class
.\"	+ the name within the class or following the prefix
.de Sj
.ie '\*[TYPE]'JAVA'\{\
.TP 5
Db.\\$1\}
.el\{\
.TP 5
\\$1\}
..
.\" The general information text macro.
.de Gn
.ie '\*[TYPE]'C'\{The DB library is a family of groups of functions that provides a modular
programming interface to transactions and record-oriented file access.
The library includes support for transactions, locking, logging and file
page caching, as well as various indexed access methods.
Many of the functional groups (e.g., the file page caching functions)
are useful independent of the other DB functions,
although some functional groups are explicitly based on other functional
groups (e.g., transactions and logging).
\}
.el\{The DB library is a family of classes that provides a modular
programming interface to transactions and record-oriented file access.
The library includes support for transactions, locking, logging and file
page caching, as well as various indexed access methods.
Many of the classes (e.g., the file page caching class)
are useful independent of the other DB classes,
although some classes are explicitly based on other classes
(e.g., transactions and logging).
\}
For a general description of the DB package, see
.IR db_intro (3).
..
.\" The library error macro, the local error macro.
.\" These macros take one argument:
.\"	+ the function name.
.de Ee
The
.I \\$1
.ie '\*[TYPE]'C'\{function may fail and return
.I errno
\}
.el\{method may fail and throw a
.IR DbException (3)
.if '\*[TYPE]'CXX'\{
or return
.I errno
\}
\}
for any of the errors specified for the following DB and library functions:
..
.de Ec
In addition, the
.I \\$1
.ie '\*[TYPE]'C'\{function may fail and return
.I errno
\}
.el\{method may fail and throw a
.IR DbException (3)
.ie '\*[TYPE]'CXX'\{or return
.I errno
\}
.el\{encapsulating an
.I errno
\}
\}
for the following conditions:
..
.de Ea
[EAGAIN]
A lock was unavailable.
..
.de Eb
[EBUSY]
The shared memory region was in use and the force flag was not set.
..
.de Em
[EAGAIN]
The shared memory region was locked and (repeatedly) unavailable.
..
.de Ei
[EINVAL]
An invalid flag value or parameter was specified.
..
.de Es
[EACCES]
An attempt was made to modify a read-only database.
..
.de Et
The DB_THREAD flag was specified and spinlocks are not implemented for
this architecture.
..
.de Ep
[EPERM]
Database corruption was detected.
All subsequent database calls (other than
.ie '\*[TYPE]'C'\{\
.IR DB->close )
\}
.el\{\
.IR Db::close )
\}
will return EPERM.
..
.de Ek
.if '\*[TYPE]'CXX'\{\
Methods marked as returning
.I errno
will, by default, throw an exception that encapsulates the error information.
The default error behavior can be changed, see
.IR DbException (3).
\}
..
.\" The SEE ALSO text macro
.de Sa
.\" make the line long for nroff.
.if n .ll 72
.nh
.na
.IR db_archive (1),
.IR db_checkpoint (1),
.IR db_deadlock (1),
.IR db_dump (1),
.IR db_load (1),
.IR db_recover (1),
.IR db_stat (1),
.IR db_intro (3),
.ie '\*[TYPE]'C'\{\
.IR db_appinit (3),
.IR db_cursor (3),
.IR db_dbm (3),
.IR db_internal (3),
.IR db_lock (3),
.IR db_log (3),
.IR db_mpool (3),
.IR db_open (3),
.IR db_thread (3),
.IR db_txn (3)
\}
.el\{\
.IR db_internal (3),
.IR db_thread (3),
.IR Db (3),
.IR Dbc (3),
.IR DbEnv (3),
.IR DbException (3),
.IR DbInfo (3),
.IR DbLock (3),
.IR DbLockTab (3),
.IR DbLog (3),
.IR DbLsn (3),
.IR DbMpool (3),
.IR DbMpoolFile (3),
.IR Dbt (3),
.IR DbTxn (3),
.IR DbTxnMgr (3)
\}
.ad
.Hy
..
.\" The function header macro.
.\" This macro takes one argument:
.\"	+ the function name.
.de Fn
.in 2
.I \\$1
.in
..
.\" The XXX_open function text macro, for merged create/open calls.
.\" This macro takes two arguments:
.\"	+ the interface, e.g., "transaction region"
.\"	+ the prefix, e.g., "txn" (or the class name for C++, e.g., "DbTxn")
.de Co
.ie '\*[TYPE]'C'\{\
.Fn \\$2_open
The
.I \\$2_open
function copies a pointer, to the \\$1 identified by the
.B directory
.IR dir ,
into the memory location referenced by
.IR regionp .
.PP
If the
.I dbenv
argument to
.I \\$2_open
was initialized using
.IR db_appinit ,
.I dir
is interpreted as described by
.IR db_appinit (3).
\}
.el\{\
.Fn \\$2::open
The
.I \\$2::open
.ie '\*[TYPE]'CXX'\{\
method copies a pointer, to the \\$1 identified by the
.B directory
.IR dir ,
into the memory location referenced by
.IR regionp .
\}
.el\{\
method returns a \\$1 identified by the
.B directory
.IR dir .
\}
.PP
If the
.I dbenv
argument to
.I \\$2::open
was initialized using
.IR DbEnv::appinit ,
.I dir
is interpreted as described by
.IR DbEnv (3).
\}
.PP
Otherwise,
if
.I dir
is not NULL,
it is interpreted relative to the current working directory of the process.
If
.I dir
is NULL,
the following environment variables are checked in order:
``TMPDIR'', ``TEMP'', and ``TMP''.
If one of them is set,
\\$1 files are created relative to the directory it specifies.
If none of them are set, the first possible one of the following
directories is used:
.IR /var/tmp ,
.IR /usr/tmp ,
.IR /temp ,
.IR /tmp ,
.I C:/temp
and
.IR C:/tmp .
.PP
All files associated with the \\$1 are created in this directory.
This directory must already exist when
.ie '\*[TYPE]'C'\{
\\$1_open
\}
.el\{\
\\$2::open
\}
is called.
If the \\$1 already exists,
the process must have permission to read and write the existing files.
If the \\$1 does not already exist,
it is optionally created and initialized.
..
.\" The common close language macro, for discarding created regions
.\" This macro takes one argument:
.\"	+ the function prefix, e.g., txn (the class name for C++, e.g., DbTxn)
.de Cc
In addition, if the
.I dir
argument to
.ie '\*[TYPE]'C'\{\
.ds Va db_appinit
.ds Vo \\$1_open
.ds Vu \\$1_unlink
\}
.el\{\
.ds Va DbEnv::appinit
.ds Vo \\$1::open
.ds Vu \\$1::unlink
\}
.I \\*(Vo
was NULL
and
.I dbenv
was not initialized using
.IR \\*(Va ,
.if '\\$1'memp'\{\
or the DB_MPOOL_PRIVATE flag was set,
\}
all files created for this shared region will be removed,
as if
.I \\*(Vu
were called.
.rm Va
.rm Vo
.rm Vu
..
.\" The DB_ENV information macro.
.\" This macro takes two arguments:
.\"	+ the function called to open, e.g., "txn_open"
.\"	+ the function called to close, e.g., "txn_close"
.de En
.ie '\*[TYPE]'C'\{\
based on the
.I dbenv
argument to
.IR \\$1 ,
which is a pointer to a structure of type DB_ENV (typedef'd in <db.h>).
It is expected that applications will use a single DB_ENV structure as the
argument to all of the subsystems in the DB package.
In order to ensure compatibility with future releases of DB, all fields of
the DB_ENV structure that are not explicitly set should be initialized to 0
before the first time the structure is used.
Do this by declaring the structure external or static, or by calling the C
library routine
.IR bzero (3)
or
.IR memset (3).
.PP
The fields of the DB_ENV structure used by
.I \\$1
are described below.
.if '\*[TYPE]'CXX'\{\
As references to the DB_ENV structure may be maintained by
.IR \\$1 ,
it is necessary that the DB_ENV structure and memory it references be valid
until the
.I \\$2
function is called.
\}
.ie '\\$1'db_appinit'\{The
.I dbenv
argument may not be NULL.
If any of the fields of the
.I dbenv
are set to 0,
defaults appropriate for the system are used where possible.
\}
.el\{If
.I dbenv
is NULL
or any of its fields are set to 0,
defaults appropriate for the system are used where possible.
\}
.PP
The following fields in the DB_ENV structure may be initialized before calling
.IR \\$1 :
\}
.el\{\
based on which set methods have been used.
It is expected that applications will use a single DbEnv object as the
argument to all of the subsystems in the DB package.
The fields of the DbEnv object used by
.I \\$1
are described below.
As references to the DbEnv object may be maintained by
.IR \\$1 ,
it is necessary that the DbEnv object and memory it references be valid
until the object is destroyed.
.ie '\\$1'appinit'\{\
The
.I dbenv
argument may not be NULL.
If any of the fields of the
.I dbenv
are set to 0,
defaults appropriate for the system are used where possible.
\}
.el\{\
Any of the DbEnv fields that are not explicitly set will default to
appropriate values.
\}
.PP
The following fields in the DbEnv object may be initialized, using the
appropriate set method, before calling
.IR \\$1 :
\}
..
.\" The DB_ENV common fields macros.
.de Se
.ie '\*[TYPE]'JAVA'\{\
.TP 5
DbErrcall db_errcall;
.ns
.TP 5
String db_errpfx;
.ns
.TP 5
int db_verbose;
The error fields of the DbEnv behave as described for
.IR DbEnv (3).
\}
.el\{\
.ie '\*[TYPE]'CXX'\{\
.TP 5
void *(*db_errcall)(char *db_errpfx, char *buffer);
.ns
.TP 5
FILE *db_errfile;
.ns
.TP 5
const char *db_errpfx;
.ns
.TP 5
class ostream *db_error_stream;
.ns
.TP 5
int db_verbose;
The error fields of the DbEnv behave as described for
.IR DbEnv (3).
\}
.el\{\
void *(*db_errcall)(char *db_errpfx, char *buffer);
.ns
.TP 5
FILE *db_errfile;
.ns
.TP 5
const char *db_errpfx;
.ns
.TP 5
int db_verbose;
The error fields of the DB_ENV behave as described for
.IR db_appinit (3).
\}
\}
..
.\" The open flags.
.de Fm
The
.I flags
and
.I mode
arguments specify how files will be opened and/or created when they
don't already exist.
The flags value is specified by
.BR or 'ing
together one or more of the following values:
.Sj DB_CREATE
Create any underlying files, as necessary.
If the files do not already exist and the DB_CREATE flag is not specified,
the call will fail.
..
.\" DB_THREAD open flag macro.
.\" This macro takes two arguments:
.\"	+ the open function name
.\"	+ the object it returns.
.de Ft
.TP 5
.Sj DB_THREAD
Cause the \\$2 handle returned by the
.I \\$1
.Al function method
to be useable by multiple threads within a single address space,
i.e., to be ``free-threaded''.
.if '\*[TYPE]'JAVA'\{\
Threading is assumed in the Java API,
so no special flags are required,
and DB functions will always behave as if the DB_THREAD flag was specified.
\}
..
.\" The mode macro.
.\" This macro takes one argument:
.\"	+ the subsystem name.
.de Mo
All files created by the \\$1 are created with mode
.I mode
(as described in
.IR chmod (2))
and modified by the process' umask value at the time of creation (see
.IR umask (2)).
The group ownership of created files is based on the system and directory
defaults, and is not further specified by DB.
..
.\" The application exits macro.
.\" This macro takes one argument:
.\"	+ the application name.
.de Ex
The
.I \\$1
utility exits 0 on success, and >0 if an error occurs.
..
.\" The application -h section.
.\" This macro takes one argument:
.\"	+ the application name
.de Dh
DB_HOME
If the
.B \-h
option is not specified and the environment variable
.I DB_HOME
is set, it is used as the path of the database home, as described in
.IR db_appinit (3).
..
.\" The function DB_HOME ENVIRONMENT VARIABLES section.
.\" This macro takes one argument:
.\"	+ the open function name
.de Eh
DB_HOME
If the
.I dbenv
argument to
.I \\$1
was initialized using
.IR db_appinit ,
the environment variable DB_HOME may be used as the path of the database
home for the interpretation of the
.I dir
argument to
.IR \\$1 ,
as described in
.IR db_appinit (3).
.if \\n(.$>1 \{Specifically,
.I \\$1
is affected by the configuration string value of \\$2.\}
..
.\" The function TMPDIR ENVIRONMENT VARIABLES section.
.\" This macro takes two arguments:
.\"	+ the interface, e.g., "transaction region"
.\"	+ the prefix, e.g., "txn" (or the class name for C++, e.g., "DbTxn")
.de Ev
TMPDIR
If the
.I dbenv
argument to
.ie '\*[TYPE]'C'\{\
.ds Vo \\$2_open
\}
.el\{\
.ds Vo \\$2::open
\}
.I \\*(Vo
was NULL or not initialized using
.IR db_appinit ,
the environment variable TMPDIR may be used as the directory in which to
create the \\$1,
as described in the
.I \\*(Vo
section above.
.rm Vo
..
.\" The unused flags macro.
.de Fl
The
.I flags
parameter is currently unused, and must be set to 0.
..
.\" The no-space TP macro.
.de Nt
.br
.ns
.TP 5
..
.\" The return values of the functions macros.
.\" Rc is the standard two-value return with a suffix for more values.
.\" Ro is the standard two-value return but there were previous values.
.\" Rt is the standard two-value return, returning errno, 0, or < 0.
.\" These macros take one argument:
.\"	+ the routine name
.de Rc
The
.I \\$1
.ie '\*[TYPE]'C'\{function returns the value of
.I errno
on failure,
0 on success,
\}
.el\{method throws a
.IR DbException (3)
.ie '\*[TYPE]'CXX'\{or returns the value of
.I errno
on failure,
0 on success,
\}
.el\{that encapsulates an
.I errno
on failure,
\}
\}
..
.de Ro
Otherwise, the
.I \\$1
.ie '\*[TYPE]'C'\{function returns the value of
.I errno
on failure and 0 on success.
\}
.el\{method throws a
.IR DbException (3)
.ie '\*[TYPE]'CXX'\{or returns the value of
.I errno
on failure and 0 on success.
\}
.el\{that encapsulates an
.I errno
on failure,
\}
\}
..
.de Rt
The
.I \\$1
.ie '\*[TYPE]'C'\{function returns the value of
.I errno
on failure and 0 on success.
\}
.el\{method throws a
.IR DbException (3)
.ie '\*[TYPE]'CXX'\{or returns the value of
.I errno
on failure and 0 on success.
\}
.el\{that encapsulates an
.I errno
on failure.
\}
\}
..
.\" The TXN id macro.
.de Tx
.IP
If the file is being accessed under transaction protection,
the
.I txnid
parameter is a transaction ID returned from
.IR txn_begin ,
otherwise, NULL.
..
.\" The XXX_unlink function text macro.
.\" This macro takes two arguments:
.\"	+ the interface, e.g., "transaction region"
.\"	+ the prefix (for C++, this is the class name)
.de Un
.ie '\*[TYPE]'C'\{\
.ds Va db_appinit
.ds Vc \\$2_close
.ds Vo \\$2_open
.ds Vu \\$2_unlink
\}
.el\{\
.ds Va DbEnv::appinit
.ds Vc \\$2::close
.ds Vo \\$2::open
.ds Vu \\$2::unlink
\}
.Fn \\*(Vu
The
.I \\*(Vu
.Al function method
destroys the \\$1 identified by the directory
.IR dir ,
removing all files used to implement the \\$1.
.ie '\\$2'log' \{(The log files themselves and the directory
.I dir
are not removed.)\}
.el \{(The directory
.I dir
is not removed.)\}
If there are processes that have called
.I \\*(Vo
without calling
.I \\*(Vc
(i.e., there are processes currently using the \\$1),
.I \\*(Vu
will fail without further action,
unless the force flag is set,
in which case
.I \\*(Vu
will attempt to remove the \\$1 files regardless of any processes
still using the \\$1.
.PP
The result of attempting to forcibly destroy the region when a process
has the region open is unspecified.
Processes using a shared memory region maintain an open file descriptor
for it.
On UNIX systems, the region removal should succeed
and processes that have already joined the region should continue to
run in the region without change,
however processes attempting to join the \\$1 will either fail or
attempt to create a new region.
On other systems, e.g., WNT, where the
.IR unlink (2)
system call will fail if any process has an open file descriptor
for the file,
the region removal will fail.
.PP
In the case of catastrophic or system failure,
database recovery must be performed (see
.IR db_recover (1)
or the DB_RECOVER flags to
.IR \\*(Va (3)).
Alternatively, if recovery is not required because no database state is
maintained across failures,
it is possible to clean up a \\$1 by removing all of the
files in the directory specified to the
.I \\*(Vo
.Al function, method,
as \\$1 files are never created in any directory other than the one
specified to
.IR \\*(Vo .
Note, however,
that this has the potential to remove files created by the other DB
subsystems in this database environment.
.PP
.Rt \\*(Vu
.rm Va
.rm Vo
.rm Vu
.rm Vc
..
.\" Signal paragraph for standard utilities.
.\" This macro takes one argument:
.\"	+ the utility name.
.de Si
The
.I \\$1
utility attaches to DB shared memory regions.
In order to avoid region corruption,
it should always be given the chance to detach and exit gracefully.
To cause
.I \\$1
to clean up after itself and exit,
send it an interrupt signal (SIGINT).
..
.\" Logging paragraph for standard utilities.
.\" This macro takes one argument:
.\"	+ the utility name.
.de Pi
.B \-L
Log the execution of the \\$1 utility to the specified file in the
following format, where ``###'' is the process ID, and the date is
the time the utility starting running.
.sp
\\$1: ### Wed Jun 15 01:23:45 EDT 1995
.sp
This file will be removed if the \\$1 utility exits gracefully.
..
.\" Malloc paragraph.
.\" This macro takes one argument:
.\"	+ the allocated object
.de Ma
.if !'\*[TYPE]'JAVA'\{\
\\$1 are created in allocated memory.
If
.I db_malloc
is non-NULL,
it is called to allocate the memory,
otherwise,
the library function
.IR malloc (3)
is used.
The function
.I db_malloc
must match the calling conventions of the
.IR malloc (3)
library routine.
Regardless,
the caller is responsible for deallocating the returned memory.
To deallocate the returned memory,
free each returned memory pointer;
pointers inside the memory do not need to be individually freed.
\}
..
.\" Underlying function paragraph.
.\" This macro takes two arguments:
.\"	+ the function name
.\"	+ the utility name
.de Uf
The
.I \\$1
.Al function method
is the underlying function used by the
.IR \\$2 (1)
utility.
See the source code for the
.I \\$2
utility for an example of using
.I \\$1
in a UNIX environment.
..
.\" Underlying function paragraph, for C++.
.\" This macro takes three arguments:
.\"	+ the C++ method name
.\"	+ the function name for C
.\"	+ the utility name
.de Ux
The
.I \\$1
method is based on the C
.I \\$2
function, which
is the underlying function used by the
.IR \\$3 (1)
utility.
See the source code for the
.I \\$3
utility for an example of using
.I \\$2
in a UNIX environment.
..
.\" Stat field macro
.de Sf
.ie '\*[TYPE]'CXX'\{\
u_int32_t \\$1;\}
.el\{\
int get_\\$1();\}
..
.TH Db 3 "January 18, 1998"
.UC 7
.SH NAME
Db \- database access class
.SH SYNOPSIS
.nf
.ft B
.ie '\*[TYPE]'CXX'\{
#include <db_cxx.h>

static int
Db::open(const char *fname, DBTYPE type,
.ti +5
int flags, int mode, DbEnv *dbenv, DbInfo *dbinfo, Db **dbpp);

DBTYPE
Db::get_type(void) const;

int
Db::close(int flags);

int
Db::cursor(DbTxn *txnid, Dbc **cursorp);

int
Db::del(Dbt *key, DbTxn *txnid);

int
Db::fd(int *fdp);

int
Db::get(DbTxn *txnid, Dbt *key, Dbt *data, int flags);

int
Db::put(DbTxn *txnid, Dbt *key, Dbt *data, int flags);

int
Db::stat(void *sp, void *(*db_malloc)(size_t), int flags);

int
Db::sync(int flags);
\}
.el\{\
import com.sleepycat.db.*;

public static Db open(
.ti +5
String fname, int type,
.ti +5
int flags, int mode, DbEnv dbenv, DbInfo dbinfo)
throws DbException;

public int get_type();

public void close(int flags)
.ti +5
throws DbException;

public Dbc cursor(DbTxn txnid)
.ti +5
throws DbException;

public void del(Dbt key, DbTxn txnid)
.ti +5
throws DbException;

public int fd()
.ti +5
throws DbException;

public void get(DbTxn txnid, Dbt key, Dbt data, int flags)
.ti +5
throws DbException;

public void put(DbTxn txnid, Dbt key, Dbt data, int flags)
.ti +5
throws DbException;

public void sync(int flags)
.ti +5
throws DbException;
\}
.ft R
.fi
.SH DESCRIPTION
.Gn
.PP
This manual page describes the Db class, which is the center of
access activity.
.PP
The currently supported file formats are btree, hashed and recno.
The btree format is a representation of a sorted, balanced tree structure.
The hashed format is an extensible, dynamic hashing scheme.
The recno format supports fixed or variable length records (optionally
retrieved from a flat text file).
.PP
Storage and retrieval for the Db access methods are based on key/data pairs,
using the Dbt class.
See
.IR Dbt (3)
for specific information on the structure and capabilities of a Dbt.
.PP
The
.I Db::open
method opens the database represented by
.I file
for both reading and writing.
Files never intended to be shared or preserved on disk may be created by
setting the file parameter to NULL.
.PP
The
.I Db::open
method
.ie '\*[TYPE]'CXX'\{\
copies a pointer to a Db object into the memory location referenced by
.IR dbpp .
\}
.el\{\
returns a Db object.
\}
The methods of this object allow you to perform various database actions,
as described below.
.Rt Db::open
.PP
Note, while most of the access methods use
.I file
as the name of an underlying file on disk,
this is not guaranteed.
Also,
calling
.I Db::open
is a reasonably expensive operation.
(This is based on a model where the DBMS keeps a set of files open for a
long time rather than opening and closing them on each query.)
.PP
The
.I type
argument is of type
.ie '\*[TYPE]'CXX'\{DBTYPE (as defined in the <db_cxx.h> include file)
and must be set to one of DB_BTREE, DB_HASH, DB_RECNO or DB_UNKNOWN.
\}
.el\{int
and must be set to one of Db.DB_BTREE, Db.DB_HASH, Db.DB_RECNO or Db.DB_UNKNOWN.
\}
If
.I type
is DB_UNKNOWN,
the database must already exist and
.I Db::open
will then determine if it is of type DB_BTREE, DB_HASH or DB_RECNO.
.PP
.Fm
.TP 5
.Sj DB_NOMMAP
Do not map this file (see
.IR DbMpool (3)
for further information).
.TP 5
.Sj DB_RDONLY
Open the database for reading only.
Any attempt to write the database using the access methods will fail
regardless of the actual permissions of any underlying files.
.Ft Db::open Db
.TP 5
.Sj DB_TRUNCATE
``Truncate'' the database if it exists, i.e.,
behave as if the database were just created,
discarding any previous contents.
.PP
.Mo "access methods"
.PP
See
.IR DbEnv (3)
for a description of the
.I dbenv
argument, and
.IR DbInfo (3)
for a description of the
.I dbinfo
argument.
.SH "Db OPERATIONS"
The Db object returned by
.I Db::open
describes a database type,
and includes a set of functions to perform various actions,
as described below.
The methods for Db are as follows:
.ie '\*[TYPE]'CXX'\{
.TP 5
DBTYPE Db::get_type(void);\}
.el\{

perl.BerkeleyDB/Changes
Revision history for Perl extension BerkeleyDB.

0.01  23 October 1997
	* first alpha release as BerkDB.

0.02  30 October 1997
	* renamed module to BerkeleyDB	
	* fixed a few bugs & added more tests

0.03  5 May 1998
	* fixed db_get with DB_SET_RECNO
	* fixed c_get with DB_SET_RECNO and DB_GET_RECNO
	* implemented BerkeleyDB::Unknown
	* implemented BerkeleyDB::Recno, including push, pop etc
	  modified the txn support. 

0.04  19 May 1998
	* Define DEFSV & SAVE_DEFSV if not already defined. This allows
	  the module to be built with Perl 5.004_04.

0.05  9 November 1998
	* Added a note to README about how to build Berkeley DB 2.x
	  when using HP-UX.
	* Minor modifications to get the module to build with DB 2.5.x

0.06  19 December 1998
	* Minor modifications to get the module to build with DB 2.6.x
	* Added support for DB 2.6.x's  Concurrent Access Method, DB_INIT_CDB.

0.07  21st September 1999
	* Numerous small bug fixes.
	* Added support for sorting duplicate values DB_DUPSORT.
	* Added support for DB_GET_BOTH & DB_NEXT_DUP.
	* Added get_dup (from DB_File).
	* beefed up the documentation.
	* Forgot to add the DB_INIT_CDB in BerkeleyDB.pm in previous release.
	* Merged the DBM Filter code from DB_File into BerkeleyDB.
	* Fixed a nasty bug where a closed transaction was still used with
	  with dp_put, db_get etc.
	* Added logic to gracefully close everything whenever a fatal error
	  happens. Previously the plug was just pulled.
	* It is now a fatal error to explicitly close an environment if there
	  is still an open database; a database when there are open cursors or
	  an open transaction; and a cursor if there is an open transaction.
	  Using object destruction doesn't have this issue, as object
	  references will ensure everything gets closed in the correct order.
	* The BOOT code now checks that the version of db.h & libdb are the
	  same - this seems to be a common problem on Linux.
	* MLDBM support added.  
	* Support for the new join cursor added.
	* Builds with Berkeley DB 3.x
   	* Updated dbinfo for Berkeley DB 3.x file formats.
	* Deprecated the TxnMgr class. As with Berkeley DB version 3,
	  txn_begin etc are now accessed via the environment object.
	
0.08  28nd November 1999
	* More documentation updates
	* Changed reference to files in /tmp in examples.t
	* Fixed a typo in softCrash that caused problems when building
	  with a thread-enabled Perl.
	* BerkeleyDB::Error wasn't initialised properly.
	* ANSI-ified all the static C functions in BerkeleyDB.xs
	* Added support for the following DB 3.x features:
	    + The Queue database type
	    + db_remove
	    + subdatabases 
	    + db_stat for Hash & Queue

0.09  29th November 1999
        * the queue.t & subdb.t test harnesses were outputting a few
          spurious warnings. This has been fixed.

0.10  8th December 1999
	* The DESTROY method was missing for BerkeleyDB::Env. This resulted in
	  a memory leak. Fixed.
	* If opening an environment or database failed, there was a small 
	  memory leak. This has been fixed. 
	* A thread-enabled Perl it could core when a database was closed. 
	  Problem traced to the strdup function.

0.11  4th June 2000
	* When built with Berkeley Db 3.x there can ba a clash with the close
	  macro.
	* Typo in the definition of DB_WRITECURSOR
	* The flags parameter wasn't getting sent to db_cursor
	* Plugged small mempory leak in db_cursor (DESTROY wasn't freeing
	  memory)
	* Can be built with Berkeley DB 3.1
   


examples/c/csv/db.c
/*-
 * See the file LICENSE for redistribution information.
 *
 * Copyright (c) 2005, 2012 Oracle and/or its affiliates.  All rights reserved.
 *
 * $Id$
 */

#include "csv.h"
#include "csv_local.h"
#include "csv_extern.h"

static int compare_uint32(DB *, const DBT *, const DBT *);

/*
 * csv_env_init --
 *	Initialize the database environment.
 */
int
csv_env_open(const char *home, int is_rdonly)
{
	int ret;

	dbenv = NULL;
	db = NULL;

	/* Create a database environment handle. */
	if ((ret = db_env_create(&dbenv, 0)) != 0) {
		fprintf(stderr,
		    "%s: db_env_create: %s\n", progname, db_strerror(ret));
		return (1);
	}

	/*
	 * Configure Berkeley DB error reporting to stderr, with our program
	 * name as the prefix.
	 */
	dbenv->set_errfile(dbenv, stderr);
	dbenv->set_errpfx(dbenv, progname);

	/*
	 * The default Berkeley DB cache size is fairly small; configure a
	 * 1MB cache for now.  This value will require tuning in the future.
	 */
	if ((ret = dbenv->set_cachesize(dbenv, 0, 1048576, 1)) != 0) {
		dbenv->err(dbenv, ret, "DB_ENV->set_cachesize");
		return (1);
	}

	/*
	 * We may be working with an existing environment -- try and join it.
	 * If that fails, create a new database environment; for now, we only
	 * need a cache, no logging, locking, or transactions.
	 */
	if ((ret = dbenv->open(dbenv, home,
	    DB_JOINENV | DB_USE_ENVIRON, 0)) != 0 &&
	    (ret = dbenv->open(dbenv, home,
	    DB_CREATE | DB_INIT_MPOOL | DB_PRIVATE | DB_USE_ENVIRON, 0)) != 0) {
		dbenv->err(dbenv, ret, "DB_ENV->open");
		return (1);
	}

	/* Create the primary database handle. */
	if ((ret = db_create(&db, dbenv, 0)) != 0) {
		dbenv->err(dbenv, ret, "db_create");
		return (1);
	}

	/*
	 * Records may be relatively large -- use a large page size.
	 */
	if ((ret = db->set_pagesize(db, 32 * 1024)) != 0) {
		dbenv->err(dbenv, ret, "DB->set_pagesize");
		return (1);
	}

	/*
	 * The primary database uses an integer as its key; on little-endian
	 * machines, integers sort badly using the default Berkeley DB sort
	 * function (which is lexicographic).  Specify a comparison function
	 * for the database.
	 */
	if ((ret = db->set_bt_compare(db, compare_uint32)) != 0) {
		dbenv->err(dbenv, ret, "DB->set_bt_compare");
		return (1);
	}

	/* Open the primary database. */
	if ((ret = db->open(db, NULL,
	    "primary", NULL, DB_BTREE, is_rdonly ? 0 : DB_CREATE, 0)) != 0) {
		dbenv->err(dbenv, ret, "DB->open: primary");
		return (1);
	}

	/* Open the secondaries.  */
	if ((ret = csv_secondary_open()) != 0)
		return (1);

	return (0);
}

/*
 * csv_env_close --
 *	Discard the database environment.
 */
int
csv_env_close()
{
	int ret, t_ret;

	ret = 0;

	/* Close the secondaries. */
	ret = csv_secondary_close();

	/* Close the primary handle. */
	if (db != NULL && (t_ret = db->close(db, 0)) != 0) {
		dbenv->err(dbenv, ret, "DB->close");
		if (ret == 0)
			ret = t_ret;
	}
	if ((t_ret = dbenv->close(dbenv, 0)) != 0) {
		fprintf(stderr,
		    "%s: DB_ENV->close: %s\n", progname, db_strerror(ret));
		if (ret == 0)
			ret = t_ret;
	}

	return (ret);
}

/*
 * csv_secondary_open --
 *	Open any secondary indices.
 */
int
csv_secondary_open()
{
	DB *sdb;
	DbField *f;
	int ret, (*fcmp)(DB *, const DBT *, const DBT *);

	/*
	 * Create secondary database handles.
	 */
	for (f = fieldlist; f->name != NULL; ++f) {
		if (f->indx == 0)
			continue;

		if ((ret = db_create(&sdb, dbenv, 0)) != 0) {
			dbenv->err(dbenv, ret, "db_create");
			return (1);
		}
		sdb->app_private = f;

		/* Keys are small, use a relatively small page size. */
		if ((ret = sdb->set_pagesize(sdb, 8 * 1024)) != 0) {
			dbenv->err(dbenv, ret, "DB->set_pagesize");
			return (1);
		}

		/*
		 * Sort the database based on the underlying type.  Skip
		 * strings, Berkeley DB defaults to lexicographic sort.
		 */
		switch (f->type) {
		case DOUBLE:
			fcmp = compare_double;
			break;
		case UNSIGNED_LONG:
			fcmp = compare_ulong;
			break;
		case NOTSET:
		case STRING:
		default:
			fcmp = NULL;
			break;
		}
		if (fcmp != NULL &&
		    (ret = sdb->set_bt_compare(sdb, fcmp)) != 0) {
			dbenv->err(dbenv, ret, "DB->set_bt_compare");
			return (1);
		}

		/* Always configure secondaries for sorted duplicates. */
		if ((ret = sdb->set_flags(sdb, DB_DUPSORT)) != 0) {
			dbenv->err(dbenv, ret, "DB->set_flags");
			return (1);
		}
		if ((ret = sdb->set_dup_compare(sdb, compare_ulong)) != 0) {
			dbenv->err(dbenv, ret, "DB->set_dup_compare");
			return (1);
		}

		if ((ret = sdb->open(
		    sdb, NULL, f->name, NULL, DB_BTREE, DB_CREATE, 0)) != 0) {
			dbenv->err(dbenv, ret, "DB->open: %s", f->name);
			return (1);
		}
		if ((ret = sdb->associate(
		    db, NULL, sdb, secondary_callback, DB_CREATE)) != 0) {
			dbenv->err(dbenv, ret, "DB->set_associate");
			return (1);
		}
		f->secondary = sdb;
	}

	return (0);
}

/*
 * csv_secondary_close --
 *	Close any secondary indices.
 */
int
csv_secondary_close()
{
	DbField *f;
	int ret, t_ret;

	ret = 0;
	for (f = fieldlist; f->name != NULL; ++f)
		if (f->secondary != NULL && (t_ret =
		    f->secondary->close(f->secondary, 0)) != 0 && ret == 0)
			ret = t_ret;

	return (ret);
}

/*
 * compare_uint32 --
 *	Compare two keys.
 */
static int
compare_uint32(DB *db_arg, const DBT *a_arg, const DBT *b_arg)
{
	u_int32_t a, b;

	db_arg = db_arg;			/* Quiet compiler. */

	memcpy(&a, a_arg->data, sizeof(u_int32_t));
	memcpy(&b, b_arg->data, sizeof(u_int32_t));
	return (a > b ? 1 : ((a < b) ? -1 : 0));
}
